### DISCUSSION TITLE
How do you initialize ecal in a class?

### DISCUSSION THREAD

**Discussion by rlewkowicz** [eclipse-ecal/ecal#1517] [General]
I'm not a great c++ dev, I used ecal about 2 years ago and am playing with it now. I remember this last time. Before I go into a main loop I have a class:
```
class Mouse
{
public:
	explicit Mouse();
	void send(int x, int y);
	eCAL::CPublisher pub;
};
```
And I want to initialize this and then use send later to publish. 

`Mouse mouse = Mouse();`

![image](https://github.com/eclipse-ecal/ecal/assets/17415445/6cca2bb8-749d-45cd-a4fd-559e2f5bd746)

I'll go dig my old code but I recall this being a pain and I had to do goofy stuff to get this to work. What is the idiomatic way to initialize ecal?

**Comment by KerstinKeller** [comment]
Hi @rlewkowicz, 
the problem in your case is, that eCAL Publishers / Subscribers cannot be copied (copy assigned / copy constructed).
This is a design decision, what happens if you copy a Publisher? Do you now have two Publishers, publishing on the same topic?

This is why we deleted the respective operators. However, this means, as the Publisher is part of your `Mouse` class, that class can't be copied either. This could be relevant if you want to store the `Mouse` in a container, but you could also construct it inside the container for example.

Better code would be:

```
Mouse mouse_1; // implicitly calls constructor
Mouse mouse_2("Mouse") // will publish on topic "Mouse"
```

**Reply by rlewkowicz** [reply]
Ty!
